Graphing json (#4033)
DataDog/documentation: content/en/graphing/graphing_json/request_json.md

---
title: Request JSON schema
kind: documentation
further_reading:
- link: "graphing/graphing_json/"
  tag: "Documentation"
  text: "Building Dashboard using JSON"
- link: "graphing/graphing_json/widget_json/"
  tag: "Documentation"
  text: "Widget JSON schema"
---

The general format for the `REQUEST_SCHEMA` is an array of one or more `requests`:

```
"requests": [
  {
    "q": "function(aggregation method:metric{scope} [by {group}])"
  }
]
```

If your `requests` parameter has multiple `requests`, the widget displays all of them:

```
"requests": [
  {
    "q": "<METRIC_1>{<SCOPE_1>}"
  },
  {
    "apm_query": "<METRIC_2>{<SCOPE_2>}"
  },
  {
    "log_query": "<METRIC_3>{<SCOPE_3>}"
  }
]
```

{{< img src="graphing/graphing_json/multi-lines.png" alt="multi lines" responsive="true" >}}

## Functions

You can apply functions to the result of each query.

> A few of these functions are further explained in a series of examples. See the [examples page][1] for more details.

Reviewer: ```suggestion
For more information about functions explained through a series of examples, see the [examples page][1].
```

#### Aggregation Method

In most cases, the number of data points available outnumbers the maximum number that can be shown on screen. To overcome this, the data is aggregated using one of 4 available methods: average, max, min, and sum.

#### Metrics

Metrics are the main focus of the graph. You can find the list of metrics available to you in the [Metrics Summary][2]. Click on any metric to see more detail about that metric, including the type of data collected, units, tags, hosts, and more.

## Scope

A scope lets you filter a series. It can be a host, a device on a host, or any arbitrary tag you can think of that contains only alphanumeric characters, plus colons and underscores (`[a-zA-Z0-9:_]+`).

Examples of scopes and their meanings:

| Scope | Meaning |
| --- | --- |
| `host:my_host` | Related to a given host. |
| `host:my_host, device:my_device` | Related to a given device on a given host. |
| `source:my_source` | Related to a given source. |
| `my_tag` | Related to a tagged group of hosts. |
| `my:tag` | Same as above. |
| `*` | Wildcard for everything. |

#### Groups

> For any given metric, data may come from a number of hosts. The data is normally aggregated from all these hosts to a single value for each time slot. You can split these out by any tag. To include a data point separated out by each host, use {host} for your group.

Reviewer: ```suggestion
For any given metric, data may come from several hosts. The data is normally aggregated from all these hosts to a single value for each time slot. You can split these out by any tag. To include a data point separated by each host, use {host} for your group.
```

#### Arithmetic

You can apply simple arithmetic to a series (+, -, * and /). 

The following example graphs 5-minute load and its double:

```json
{
  "viz": "timeseries",
  "requests": [
    {
      "q": "system.load.5{intake} * 2"
    },
    {
      "q": "system.load.5{intake}"
    }
  ]
}
```

> You can also add, substract, multiply, and divide a series. Note that Datadog does not enforce consistency at this point, so you *can* divide apples by oranges.

Reviewer: ```suggestion
You can also add, subtract, multiply, and divide a series. Note that Datadog does not enforce consistency at this point, so you *can* divide apples by oranges.
```

```json
{
  "viz": "timeseries",
  "requests": [
    {
      "q": "metric{apples} / metric{oranges}"
    }
  ]
}
```

## Stacked Series

{{< img src="graphing/graphing_json/slice-n-stack.png" alt="slice and stack" responsive="true" >}}

In the case of related timeseries, you can draw them as stacked areas by using the following syntax:

```
"requests": [
  {
    "q": "metric1{scope}, metric2{scope}, metric3{scope}"
  }
]
```

Instead of one query per chart, you can aggregate all queries into one and concatenate the queries.

## Slice-n-Stack

You can represent a metric shared across hosts and stack the results. For instance, when selecting a tag that applies to more than 1 host, you see that ingress and egress traffic is nicely stacked to give you the sum as well as the split per host. This is useful to spot wild swings in the distribution of network traffic.

Here's how to do it for any metric:

```
"requests" [
  {
    "q": "system.net.bytes_rcvd{some_tag, device:eth0} by {host}"
  }
]
```

> Note that in this case you can only have 1 query. But you can also split by device, or a combination of both:

Reviewer: ```suggestion
Note that in this case, you can only have 1 query. But you can also split by device or a combination of both:
```
